### ISSUE TITLE
Mailbox/query with 'role' filter does not work. It matches any mailbox that *doesn't* have a role!

### ISSUE THREAD

**Issue by wolfsage** [cyrusimap/cyrus-imapd#2895]
The following code is buggy, if there is a filter->role but there is *not* a mailbox role, it will never set is_match to 0 so it thinks it matched.

```
        if (JNOTNULL(filter->has_any_role) || filter->role) {
            mbname_t *mbname = mbname_from_intname(rec->mboxname);
            char *role = _mbox_get_role(query->req, mbname);
            int has_role = role != NULL;
            int is_match = 1;
            if (JNOTNULL(filter->has_any_role)) {
                is_match = (filter->has_any_role == json_true()) == (has_role != 0);
            }
            if (filter->role == json_null() && role) {
                is_match = 0;
            }
            else if (JNOTNULL(filter->role) && role) {
                if (is_match) is_match = !strcmp(json_string_value(filter->role), role);
            }
            free(role);
            mbname_free(&mbname);
          if (!is_match) return 0;
```

Fix might be:
```
       if (filter->role == json_null() && role) {
                is_match = 0;
            }
            else if (JNOTNULL(filter->role) && ! role) {
              is_match = 0;
            }
            else if (JNOTNULL(filter->role) && role) {
                if (is_match) is_match = !strcmp(json_string_value(filter->role), role);
            }
            free(role);
            mbname_free(&mbname);
          if (!is_match) return 0;
```

Or set is_match to 0 early on and only set it to 1 under clear match conditions?

**Comment by ksmurchison** [comment]
Potential fix:
https://github.com/cyrusimap/cyrus-imapd/pull/2896

**Comment by ksmurchison** [comment]
Should be fixed by #2896
